File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/test/graph.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import { Graph } from '../src/grammarGraph/graph.ts';
import {
  RuleChar,
  RuleCharExclude,
  RuleEnd,
} from '../src/grammarGraph/grammarGraphTypes.ts';
import type { UnresolvedRule } from '../src/grammarGraph/grammarGraphTypes.ts';
import { GraphNode } from '../src/grammarGraph/graphNode.ts';
import { GraphPointer } from '../src/grammarGraph/graphPointer.ts';
import { Pointers } from '../src/grammarGraph/pointers.ts';
import { RuleRef } from '../src/grammarGraph/ruleRef.ts';
import { isRule } from '../src/grammarGraph/typeGuards.ts';

describe('Graph', () => {
  const grammar = 'example-grammar';
  const stackedRules: UnresolvedRule[][][] = [
    [[new RuleChar([65, 66, 67])], [new RuleChar([68, 69, 70])]],
    [[new RuleChar([71, 72, 73])], [new RuleChar([74, 75, 76])]],
  ];
  const rootId = 0;

  it('creates a Graph instance', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    assert.ok(graph instanceof Graph);
    assert.equal(graph.grammar, grammar);
  });

  it('gets the root node', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    const rootNode = graph.getRootNode(rootId);
    assert.ok(rootNode instanceof Map);
    assert.equal(rootNode.size, 2);
  });

  it('throws when asked for a root node that does not exist', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    assert.throws(() => graph.getRootNode(99), /Root node not found for value: 99/u);
  });

  it('gets the initial pointers', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    const pointers = graph.getInitialPointers();
    assert.equal(pointers.size, 2);
    for (const pointer of pointers) {
      assert.ok(isRule(pointer.rule));
    }
  });

  it('prints the graph', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    const printedGraph = graph.print(null, true);
    assert.equal(typeof printedGraph, 'string');
    assert.ok(printedGraph.length > 0);
  });

  it('iterates over pointers', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    const rules: UnresolvedRule[] = [
      new RuleChar([65, 66, 67]),
      new RuleChar([68, 69, 70]),
      new RuleChar([71, 72, 73]),
      new RuleChar([74, 75, 76]),
      new RuleCharExclude([1]),
      new RuleEnd(),
    ];
    const mockPointers = new Pointers(
      ...rules.map(
        (rule, idx) => new GraphPointer(new GraphNode(rule, { stackId: idx, pathId: 0, stepId: 0 })),
      ),
    );

    const result = [...graph.iterateOverPointers(mockPointers)];
    assert.equal(result.length, 6);
    for (const [rule, pointers] of result) {
      assert.ok(isRule(rule));
      assert.ok(pointers.length >= 1);
      for (const pointer of pointers) {
        assert.ok([...mockPointers].includes(pointer));
      }
    }
  });

  it('raises an error on a reference rule', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    const mockPointers = [
      new GraphPointer(new GraphNode(new RuleRef(0), { stackId: 0, pathId: 0, stepId: 0 })),
    ];
    assert.throws(
      () => [...graph.iterateOverPointers(mockPointers)],
      /Encountered a reference rule in the graph/u,
    );
  });

  it('rejects non-string input in add', () => {
    const graph = new Graph(grammar, stackedRules, rootId);
    assert.throws(
      () => graph.add(65 as unknown as string),
      /src must be a string in graph\.add/u,
    );
  });
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_graph/graph_test.py
import pytest

from .grammar_graph_types import RuleChar, RuleCharExclude, RuleEnd
from .graph import Graph
from .graph_node import GraphNode, GraphNodeMeta
from .graph_pointer import GraphPointer
from .pointers import Pointers
from .rule_ref import RuleRef
from .type_guards import is_rule

GRAMMAR = "example-grammar"
ROOT_ID = 0


def stacked_rules():
    return [
        [
            [RuleChar([65, 66, 67])],
            [RuleChar([68, 69, 70])],
        ],
        [
            [RuleChar([71, 72, 73])],
            [RuleChar([74, 75, 76])],
        ],
    ]


def test_should_create_a_graph_instance():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    assert isinstance(graph, Graph)
    assert graph.grammar == GRAMMAR


def test_should_get_the_root_node():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    root_node = graph.get_root_node(ROOT_ID)
    assert isinstance(root_node, dict)
    assert len(root_node) == 2


def test_should_get_the_initial_pointers():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    pointers = graph.get_initial_pointers()
    assert isinstance(pointers, Pointers)
    assert pointers.size == 2


def test_should_print_the_graph():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    printed_graph = graph.print(None, True)
    assert isinstance(printed_graph, str)
    assert len(printed_graph) > 0


def test_should_iterate_over_pointers():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    rules = [
        RuleChar([65, 66, 67]),
        RuleChar([68, 69, 70]),
        RuleChar([71, 72, 73]),
        RuleChar([74, 75, 76]),
        RuleCharExclude([1]),
        RuleEnd(),
    ]
    mock_pointers = Pointers(
        *[
            GraphPointer(GraphNode(rule, GraphNodeMeta(idx, idx, idx)))
            for idx, rule in enumerate(rules)
        ]
    )

    result = list(graph.iterate_over_pointers(mock_pointers))
    assert len(result) == 6
    all_pointers = list(mock_pointers)
    for rule, pointers in result:
        assert is_rule(rule) is True
        assert len(pointers) >= 1
        for pointer in pointers:
            assert any(pointer is candidate for candidate in all_pointers)


def test_should_raise_error_on_reference_rule():
    graph = Graph(GRAMMAR, stacked_rules(), ROOT_ID)
    mock_pointers = [
        GraphPointer(GraphNode(RuleRef(0), GraphNodeMeta(0, 0, 0))),
    ]

    with pytest.raises(ValueError, match="Encountered a reference rule in the graph"):
        list(graph.iterate_over_pointers(mock_pointers))
